### ISSUE TITLE
Better API

### ISSUE THREAD

**Issue by Gozala** [ipfs/js-ipfs-block#47]
Reflecting a bit more on the https://github.com/ipld/interface-ipld-format/issues/52 and https://github.com/ipld/ipld/issues/64#issuecomment-473442520 I'm realizing that [IPLD Block API](https://github.com/ipld/js-ipld-block) is a primary issue for many of the issues.

As it stands now it's a glorified tuple `[cid, data]`. However there are (at least) handful of cases where you'd want encoded data but not the corresponding `cid`. In other words the way things stand today you can refer to data encoded in specific format without `CID`.

I would like to propose to rethink a `Block` as a glorified `[codec, data]` instead ? That would still allow deriving `cid` on demand, but it would also allow provide a way to refer to data in specefic format without `cid`, which is helpful to allow ["inline blocks"](https://github.com/ipld/ipld/issues/64#issuecomment-473442520) in the future.

Here what such `Block` implementation could look like:

```js
class Block {
  /*::
  data:Uint8Array
  // Registered codec
  // https://github.com/multiformats/multicodec/blob/master/table.csv
  format:string
  */
  constructor(data, format) {
    this.format = format
    this.data = data
  }
  toCID(options) {
    return Block.toCID(this, options)
  }
  static async toCID({data, format}/*:Block*/, {hashAlg, cidVersion}={})/*:CID*/ {
    const hash = await multihashes(this.data, hashAlg || defaultHashAlg)
    const version = cidVersion != null ? cidVersion : 1
    return CID(version, format, hash)
  }
  static encode(block/*:Block*/)/*:Uint8Array*/ {
    return multicodec.addPrefix(block.format, block.data)
  }
  static decode(bytes/*:Uint8Array*/)/*:Block*/ {
    return new Block(multicodec.rmPrefix(this.codec, this.data), multicodec.getCodec(bytes))
  }
}
``` 

Note: In theory `cid` could just be a getter, however it's somewhat unclear how the hashing algorithm get's selected, format is required to provide `defaultHashAlg` but I'm starting to question if choice of hashing algorithm should be codec implementors responsibility. Seems to me that consumer should be in charge of that. It also appears that use of canonical default hashing algorithm if one isn't provided is likely a reasonable option.

Another layer could ensure that `toCID()` is passed the `hashAlg` that corresponds to the `defaultHashAlg` of the block's format.

This change would also pave way to more simplifications:

- entries of the node could then be represented as union of `Block` and `CID` where `CID` is a link into other node and `Block` is an inline node in arbitrary format (implying it can be different from format from the node containing it).
- This would essentially provide a way to represent **inline nodes** discussed here https://github.com/ipld/ipld/issues/64#issuecomment-473442520
- At the moment IPLD provides interop across multiple formats as long as each block is encoded in a specific format, failing to support multiple formats with-in a single block. Which I'd argue precludes too many assumptions about data layout. This would enable composed formats discussed https://github.com/ipld/ipld/issues/63

**Comment by rvagg** [comment]
@Gozala to clarify: your primary objection the current framing is that the CID isn't always helpful or necessary so deriving it on-demand would be just as good?

Regarding "but it would also allow provide a way to refer to data in specefic format without `cid`", isn't that a job for the selector though? A good enough selector spec should be able to be mostly CID agnostic with the exception of the root. Or maybe you're seeing more complexity here than me (likely) so could you outline a use-case where the CID might get in the way?

And back to our async crypto discussion, wouldn't optional (/ lazy) CID generation be a cause for async crypto in the browser? So `toCID()` will either need to be `async` or have an already cached CID ready to serve when in the browser if it's to use any of the Web API crypto hash algorithms?

**Comment by Gozala** [comment]
> @Gozala to clarify: your primary objection the current framing is that the CID isn't always helpful or necessary so deriving it on-demand would be just as good?

> Regarding "but it would also allow provide a way to refer to data in specefic format without cid", isn't that a job for the selector though? A good enough selector spec should be able to be mostly CID agnostic with the exception of the root.

Maybe better way to explain what I mean is data can be in following states:

1. raw data
2. encoded data in a specific format
3. addressed data - CID for encoded data

At the moment 2 does really exist, or rather exists in form where you have encoded data but without knowing what the format is. And only way to make sense of that data is to also have 3 to be able to figure what the format is. This is problematic because sometimes you want to inline encoded data in a block instead of saving it separately & making it available through own CID. 

> Or maybe you're seeing more complexity here than me (likely) so could you outline a use-case where the CID might get in the way?

It just this little dance you have to do three steps forward two steps back.

> And back to our async crypto discussion, wouldn't optional (/ lazy) CID generation be a cause for async crypto in the browser? So toCID() will either need to be async or have an already cached CID ready to serve when in the browser if it's to use any of the Web API crypto hash algorithms?

This is not related to crypto, data here is assumed to be already serialized with a format.

**Comment by rvagg** [comment]
> This is not related to crypto, data here is assumed to be already serialized with a format.
yeah, you're right, I was getting hashing too mixed up with CID creation I suppose.

**Comment by vmx** [comment]
I like that proposal. A Block should be a unit so that you can make sense of that data within it. But for that you don't need to know the CID, knowing the codec is enough.

Also the idea behind the CID is that you can also use other hashing algorithms. The current IPLD code kind of expects the CID to always be the same, hence using the same hashing algorithm for the same Block.

>  but I'm starting to question if choice of hashing algorithm should be codec implementors responsibility.

The reason that there's the `defaultHashAlg` property is, that practically spoken, lots of merkle-tree based data is hard-coded to a specific algorithm (if you think about Git (although I think that changes), blockchains or even IPFS). This way the consumer wouldn't always need to remember and pass in the correct hash algorithm.

**Comment by mikeal** [comment]
> A Block should be a unit so that you can make sense of that data within it. But for that you don't need to know the CID, knowing the codec is enough. Also the idea behind the CID is that you can also use other hashing algorithms. The current IPLD code kind of expects the CID to always be the same, hence using the same hashing algorithm for the same Block.

This is a good point, but I do wonder how practical it is. You need a full unique identifier in order to:

* Share the block on the network
* Store it locally
* Cache it in an LRU
* Link to the block from another block

So basically, most operations you want to do in memory, disc and network will require the full CID.

I like the idea of exposing the codec without needing a full CID, that’s a great idea and we should do that, it would save us the trouble of computing the hash in cases where we are creating new blocks. But in the cases where we aren’t authoring new blocks but simply reading them from storage or the network we need to be able to instantiate them with a CID that won’t change, even if we offer an easy way to get a **new** CID for the same block.

Multihashes offer us an easy upgrade path to new hashing algorithms but they aren’t magic, if you change the hashing algorithm you’ve effectively “forked” the block. No existing links to the block will work and no references anyone has to it will work anymore.

> As it stands now it's a glorified tuple [cid, data]. 

Correct. It’s also worth noting that [“Block”](https://github.com/ipld/specs#block) is a general term we use across IPLD to refer to this pairing. We can make the JS implementation of Block take on whatever features we want, but the “concept” of capturing a [CID, Data] tuple has to remain intact, even if we compute the CID lazily in some cases.

> However there are (at least) handful of cases where you'd want encoded data but not the corresponding cid. 

I agree with this.

> In other words the way things stand today you can refer to data encoded in specific format without CID.

I don’t know if I agree with this. How do you “refer to data” without the full identifier?

**Comment by mikeal** [comment]
> it would save us the trouble of computing the hash in cases where we are creating new blocks

I’ll expand on this a little.

I’m working on some proof-of-concept collection stuff and it’s becoming clear than some interfaces we want to have that produce blocks may end up producing a lot of blocks we never actually use. For instance, blocks for intermediary nodes that may be over-written before they ever get stored.

@Gozala’s ideas around lazily computing the CID got me thinking that we could *also* lazily compute the encoded data as well. This would get rid of pretty much all the perf concerns with these “throwaway blocks” because if nobody ever accesses the binary data or the CID then we never take the perf hit.

**Comment by mikeal** [comment]
Another issue I see with this interface proposal is that CID creation is synchronous. Some hashing algorithms are only available async, which is why we’ve made all the multihash and cid creation stuff async in other interfaces.

Related, if we did lazy encoding would that also have to be async? Do we know of any encoders that require async?

**Comment by Gozala** [comment]
> > In other words the way things stand today you can refer to data encoded in specific format without CID.
> 
> I don’t know if I agree with this. How do you “refer to data” without the full identifier?

I think it's really hard to convey the full picture as I'm pulling on multiple threads (e.g. https://github.com/ipld/ipld/issues/67) that are interconnected.

What I meant to say today you can either get reference to:

- encoded block (bytes), which does not tells you what the format is.
- CID - which tells you format and gives you a way to get encoded block to decode it.

What I find missing is way to have reference to encoded block (bytes) that carry information about the format. You can workaround that with multicodec but there is no built-in support for it in the IPLD.

**Comment by Gozala** [comment]
> Another issue I see with this interface proposal is that CID creation is synchronous. Some hashing algorithms are only available async, which is why we’ve made all the multihash and cid creation stuff async in other interfaces.

Maybe I have overlooked something, as far as I could tell multihash was synchronous (or at least that's what docs suggest https://github.com/multiformats/js-multihash/blob/443a8ebdd145582f2cac6e708ffe890401d5de08/README.md

I have no objections to making that async, just wanted to keep API mostly compatible with what's already there (I was hoping `.cid` getter could have being also added for backwards compatibility).

> Related, if we did lazy encoding would that also have to be async? Do we know of any encoders that require async?

Yes secret-box described here (https://github.com/ipld/ipld/issues/67) is one and I also have plane to make one for SSB [private-box](https://github.com/auditdrivencrypto/private-box). Those need to be async.

**Comment by Gozala** [comment]
> @Gozala’s ideas around lazily computing the CID got me thinking that we could _also_ lazily compute the encoded data as well. This would get rid of pretty much all the perf concerns with these “throwaway blocks” because if nobody ever accesses the binary data or the CID then we never take the perf hit.

That also crossed my mind, but that implied that you would need to make decode / encode here async as well. 

Generally I think across different threads we get to the more or less same thing - There needs to be lower layer that is just about blocks as in raw bytes, and another layer that is higher level blocks that is representation of the data encoded in bytes. I think we need both and need two different terms to talk about them. I'd offer following:

- Block - raw bytes + format `[bytes, format]`
- BlockView - data model for the `Block`. 

That way you could have a lazy block as per @mikeal's proposal:

```js
// lazy block
const view = new BlockView({x:1, y:2}, 'dag-cbor')
// Block instance that encoded data
const block = await view.toBlock()

// and you can go other way round too
const view2 = await BlockView.fromBlock(block)
```

**Comment by Gozala** [comment]
In fact `BlockView` here could be more or less what @mikeal described as `Reader` in https://github.com/ipld/interface-ipld-format/issues/52

**Comment by mikeal** [comment]
> js-multihash is sync

That’s because that library doesn’t actually do the hashing, it just takes a hash you’ve already computed and sticks the multihash data in front of it. Check out https://www.npmjs.com/package/multihashing-async for a library that actually implements the hashing.

**Comment by Gozala** [comment]
This got me thinking that multihashing is the form of lossy codec (encoder to be precise), I wonder if same interface for hashing would make sense here as well.

**Comment by Gozala** [comment]
@mikeal I have started playing with an idea of `BlockView` on top of the proposed `Block` API and that got me wondering about use cases you have about lazy encoding. Specifically it seems to me that `BlockView` may not want to be tied to a format more or less how [`DataView`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) for in relation to typed arrays.

Additionally I started to wonder if `BlockView` might provide a nice abstraction for dealing with "remaining path", specifically currently we're discussing that resolve / get returns `{cid, remainingPath}  | {value}` maybe instead it could just always return `BlockView` instance and `BlockView` instance needs to have `path` with-in the target `block` more or less how [`DataView`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) can be constructed with a `ArrayBuffer` and `byteOffset, byteLength` to represent view with in that slice.

**Comment by Gozala** [comment]
To elaborate a bit more on not tying `BlockView` to the format - It seems to me that you may want to decide format you want to encode whenever you are encoding, I also imagine that one might want to encode same block data into two different formats.

**Comment by Gozala** [comment]
This is a sketch I end up with, and I'd love feedback

```js
// @flow strict

declare class Block<a> {
  +data: BinaryEncoded<a>;
  +format: Format;
  equals(mixed): boolean;
  toCID():Promise<CID<a>>;

  static constructor(
    data: BinaryEncoded<a>,
    format: Format,
    // Can be passed to override default CID hash algorithm, version for the format.
    options?: CIDOptions
  ): void;
  static toCID(Block<a>): Eventual<CID<a>>;
  static toBytes(Block<a>): BinaryEncoded<a>;
  static fromBytes(BinaryEncoded<a>): Block<a>;
}

type CIDOptions = {
  hashAlg?: string,
  cidVersion?: 0 | 1
}

declare class BlockView<a> {
  static fromBlock(source: Block<a>): BlockView<a>;
  static fromCID(source: CID<a>, path?: string): BlockView<a>;
  static constructor(a, format?: Format): BlockView<a>;

  // Returns async iterator of all the paths
  paths(): AsyncIterable<string>;
  // Returns async iterator of all the links
  links<b>(): AsyncIterable<CID<b>>;
  // Returns async iterator of all the inlined blocks
  blocks<b>(): AsyncIterable<BlockView<b>>;
  // Retruns asyn iterator for all the entries inline blocks and links to blocks
  entries<b>(): AsyncIterable<[string, BlockView<a>]>;

  // Returns view for the given path.
  get<b>(path: string): Eventual<BlockView<b>>;
  // Creates a new block view that has entry for the given path updated with
  // an inline block.
  set<b>(path: string, BlockView<b>): Eventual<BlockView<a>>;
  link<b>(path: string, CID<b>): Eventual<BlockView<a>>;
  // creates a new block view that has entry for the given path deleted.
  delete(path: string): Eventual<BlockView<a>>;
  // Edits this block and returns modified version.
  // If path is `/` (which is the default if path not passed) mutable version
  // of this view is passed and multiple changes can be done in a batch. If path
  // is something other than `/` update function is passed mutable view for the
  // entry under returned block view get's copy of this with path updated.
  edit<b>(
    update: (MutableBlockView<b>) => Promise<BlockView<b>>,
    path?: string
  ): Eventual<BlockView<a>>;

  // Encodes block with passed format, if omitted uses default which is either
  // whatever source of the view had, what was passed to the constructor or
  // canonical default: dag-cbor.
  toBlock(format?: Format): Eventual<Block<a>>;
}

declare class MutableBlockView<a> extends BlockView<a> {
  set<b>(string, BlockView<b>): Eventual<MutableBlockView<a>>;
}

declare class CID<a> {
  toString(): string;
  toBaseEncodedString(base?: string): string;
  equals(mixed): boolean;
}

// Registered codec identifier from the following table
// https://github.com/multiformats/multicodec/blob/master/table.csv
export opaque type Format: string = string
export opaque type BinaryEncoded<data>: Uint8Array = Uint8Array

type Eventual<+a> = a | Promise<a>
```

**Comment by mikeal** [comment]
This morning @vmx and I had a good chat about this.

First of all, something that is about to get rather confusing, is that “Block” is, conceptually, the lowest level primitive in the stack. However, in order to make the most useful JS API, it makes sense to move a lot of stuff into the Block API.

Volker’s idea, which I rather like, was to simplify `js-ipld-format` such that it only contains the Reader and encode/decode. Then, the Block API would actually **depend on** an `ipld-format` library that dynamically loads any required formats for encoding/decoding.

This would give us something like this, and Volker can correct me if I’m miss-interpreting anything from our discussion:

### Class Method: Block.encode(data, cidOptions)

Returns a Block instance.

### Class Method:  Block.from(binary, cid || cidOptions)

Returns a Block instance.

## Class: Block

This object is instantiated with enough information to give a consistent and immutable view of the binary data, the decoded data, **and** the CID. Even if the CID is not yet computed, the options for the CID to be generated are locked at instantiation.

This also means that the following code sample would **not** actually compute a CID.

```javascript
await Block.from(buffer).decode()
```

### block.decode

Promise that resolves to a decoded native object for the block. Every call returns a **new** version of the object (no caching).

### block.reader

Promise that resolves to a `Reader` for the block data (to be defined in `ipld-format`).

### block.cid

Promise that resolves to an instance of CID. Dynamically generated. Immutable.

### block.data

Promise that resolves to binary block data. Possibly dynamically generated.

——-

Thoughts? Any opinions of instance getters for promises rather than using methods?

**Comment by mikeal** [comment]
@Gozala my main piece of feedback is that we can’t have cid or binary data in the block mutate, so there must be a way to instantiate it with a full CID attached, even if that isn’t required, so all the methods that would return a new block instance need to support full cid as well as cid options — and once those options are set (either by user input or defaults) they need to also never mutate.

Also, I think we should leave mutation out of the scope of the Block API. If you want to mutate a block you’ll end up doing a full decode -> mutate -> re-encode to a new block. There may be codec specific ways of doing faster mutation, but those codecs can just implement those mutation interfaces as they see fit.

Other than that, I think we’re starting to close in on the same set of funcitonality.

**Comment by Gozala** [comment]
Feedback

- Don't like getters that do work under the hood as it happens as side effect to referring it. I much prefer explicit calls to function / method as it's clear that things will happen.

- Don't like `block.cid` because it removes ability from user to choose desired hashing algorithm. I think `toCID(?cidOpts)` that takes optional options is better in that regard. By default it can use the right settings.

- Why `block.data` is not just property containing a bytes ? I guess because Block.encode(data, cidOptions) does not actually does a work. Which is why I really prefer to have blocks be blocks and have something like `BlockView` that can be used to assemble, edit blocks.

- I think this conflates what I would prefer to have separate meaning block , block reader, block writer

**Comment by Gozala** [comment]
> @Gozala my main piece of feedback is that we can’t have cid or binary data in the block mutate, 

They don't block is immutable pair of `{format, encodedBytes, ?cidOptions}`, it's `BlockView` that allows mutations and it's aimed to do more or less what https://github.com/immutable-js/immutable-js does.

> so there must be a way to instantiate it with a full CID attached, even if that isn’t required, so all the methods that would return a new block instance need to support full cid as well as cid options — and once those options are set (either by user input or defaults) they need to also never mutate.

Are you talking about `BlockView` ? If so let me clarify:

-  `fromBlock(source: Block<a>): BlockView<a>` Creates view from the block, there for it knows `{format, encodedBytes, ?cidOptions}` there is no but it can be computed on demand. There is no attached CID because only reason you'd want to create view from block is to update it (in immutable fashion) and there for you'd be generating CID on demand for resulting block.

- `fromCID(source: CID<a>, path?: string): BlockView<a>` This has CID, however idea is here is that function will actually fetch a block and create view from it (probably lazily). After immutable changes you'd still be generating new block / cid on demand.

- `static constructor(a, format?: Format): BlockView<a>` This also mainly to let you assemble node once it's ready you generate block / cid on demand.

**Comment by Gozala** [comment]
> Also, I think we should leave mutation out of the scope of the Block API. If you want to mutate a block you’ll end up doing a full decode -> mutate -> re-encode to a new block. There may be codec specific ways of doing faster mutation, 

Sure we can leave it off. However I disagree that you would need to do it, API is designed such that `BlockView` is essentially `{source:CID|Block, changes:{path:Block|CID}`. That allows you to traverse view without materializing the block and `toBlock(format)` is what materializes it on demand. Meaning sure it will need to `decode -> mutate -> encode` but it only happens when requested.

> but those codecs can just implement those mutation interfaces as they see fit.

One of the primary benefits of IPLD to me is that it provides a canonical interface for working with data across arbitrary formats. If I need to use format specific API per block that benefit is lost. Which is why I'm proposing canonical interface in form of `BlockView` that still means codec authors can provide additional interfaces that are tailored to them.

**Comment by Gozala** [comment]
To clarify in my vision

- `BlockView` is intentionally is agnostic of underlying format / codec or CID that it will have. That is because it want't to allow user to materialize block in arbitrary format supported by the IPLD.
- `Block` is intentionally doesn't have `CID` because one can be computed on demand, and choice of hashing algorithm is still open even though by default it will follow prescription of the codec / format.
- `BlockView` allows generic functional mutation API, that is it doesn't mutate underlying block (which may not even be computed yet) but rather creates an altered view from which block in arbitrary format can be computed (It can also use internal caching to optimize block computations and even access of paths).

**Comment by mikeal** [comment]
> Don't like getters that do work under the hood as it happens as side effect to referring it. I much prefer explicit calls to function / method as it's clear that things will happen.

Ya, the more I think about it the more I like just having these be explicit method calls that implement the caching internally.

> Don't like block.cid because it removes ability from user to choose desired hashing algorithm. I think toCID(?cidOpts) that takes optional options is better in that regard. By default it can use the right settings.

You can still specify the desired hashing algorithm during instantiation instead of passing in an explicit CID. However, you **cannot** mutate the CID in any way after instantiation. This is very important, a block is assumed to be an immutable structure, if I hand you a block that I resolved through a CID reference you should not be able to mutate that CID because it’s effectively a new block as the link is now broken.

> Why block.data is not just property containing a bytes ? I guess because Block.encode(data, cidOptions) does not actually does a work. Which is why I really prefer to have blocks be blocks and have something like BlockView that can be used to assemble, edit blocks.

Correct, this is in-line with the rest of the laziness in the structure. I would be more inclined to have a separate object like BlockView if the rest of the properties didn’t already have the exact same lazy nature.

> They don't block is immutable pair of {format, encodedBytes, ?cidOptions}, it's BlockView that allows mutations and it's aimed to do more or less what https://github.com/immutable-js/immutable-js does.

Perhaps I read the API wrong, it looks like I can get a CID out of it after it’s instantiated with a different algorithm. I’d consider that a mutation of the CID.

> One of the primary benefits of IPLD to me is that it provides a canonical interface for working with data across arbitrary formats. If I need to use format specific API per block that benefit is lost. Which is why I'm proposing canonical interface in form of BlockView that still means codec authors can provide additional interfaces that are tailored to them.

I can see where you’re coming from, but all of our current codecs would need to do a full decode -> mutate -> encode anyway.

However, I **do** think we need to build a generic mutation API that works across codecs, I just think that it should do full graph manipulation rather than just block level mutation. This is where most of the pain is right now anyway, rewriting all the intermediary nodes and bubbling them up, batching many transactions together for performance, etc.

> Block is intentionally doesn't have CID because one can be computed on demand, and choice of hashing algorithm is still open even though by default it will follow prescription of the codec / format.

This is I think where we are primarily diverging. Changing the hashing algorithm is effectively block mutation. No references to that block will persist through that mutation, so as far as the graph, local storage, network content discovery, and caches are concerned, it’s the same thing as mutating the underlying data.

That is not to say you **must** instantiate with a CID attached, we could compute one lazily, but we need to ensure the options required for that hash are set at instantiation and cannot mutate.

**Comment by Gozala** [comment]
> Perhaps I read the API wrong, it looks like I can get a CID out of it after it’s instantiated with a different algorithm. I’d consider that a mutation of the CID.

But once you mutate the view (which you can't do you get new view) than it represents different block so it should have different CID.

**Comment by Gozala** [comment]
> I can see where you’re coming from, but all of our current codecs would need to do a full decode -> mutate -> encode anyway.

Sure you will once block is materialized but it's not until you call `toBlock` and when you do that's what you would have to do anyway.

Think of `BlockView` as a pointer to the original block (if there is one) and changes. You're get operations either read from aggregated changes and if not there then will get from the original block. It will only do full decode if there are changes and only when you call `toBlock` so that changes could actually be applied and encoded into a block.

**Comment by Gozala** [comment]
> However, I **do** think we need to build a generic mutation API that works across codecs, I just think that it should do full graph manipulation rather than just block level mutation. This is where most of the pain is right now anyway, rewriting all the intermediary nodes and bubbling them up, batching many transactions together for performance, etc.

Proposed API attempts to address exactly that. Each `BlockView` is a graph as well if you consider all the links it contains and mutations pretty much create batch of OPs that need to be performed to materialize an altered graph.

**Comment by Gozala** [comment]
> This is I think where we are primarily diverging. Changing the hashing algorithm is effectively block mutation. No references to that block will persist through that mutation, so as far as the graph, local storage, network content discovery, and caches are concerned, it’s the same thing as mutating the underlying data.

> That is not to say you must instantiate with a CID attached, we could compute one lazily, but we need to ensure the options required for that hash are set at instantiation and cannot mutate.

I am starting to understand your point on this, I do need little bit more info to be fully convinced though.

I do understand that generating CID with different hashing algorithm for the same block will result in different block and if that happens by mistake that would be bad, there for I can buy into argument that calling `new Block(block.data, block.format, {hashAlg:diffHashAlg})` algorithm might be a good way to prevent such mistakes.

I do however don't necessarily agree that it's full mutation because all the links from that block will be preserved (at least that's what I had in mind) there for it's more like you create new head for the mostly same graph.

So if you're suggesting that risk of accidental CID change has high risk consequences and there for API should be adjusted to reflect that I'm totally onboard. If there is something else I'm missing please elaborate.

**Comment by mikeal** [comment]
> So if you're suggesting that risk of accidental CID change has high risk consequences and there for API should be adjusted to reflect that I'm totally onboard. If there is something else I'm missing please elaborate.

Yes, this is my primary concern. If I traverse a graph with a selector and hand you a block instance, and then you hand that block to someone else’s code and they ask for a CID with their favorite algorithm they don’t have a clear indication that they’re now linking a different block than the one that was handed to them. They also don’t have a clear indication that they would need to store the block again with that CID in order for it to persist if the selector engine gave it to them from an existing storage backend.

We have these cool multihashes, and there are very good reasons we have them, but it’s not always clear to programmers that two multihashes for the same data with different algorithms are **so** different. If I wasn’t already knowledgeable about this stack and I saw that feature I’d say “great! I get to use my favorite algorithm instead of sha-whateva is the default.”

**Comment by Gozala** [comment]
@mikeal Fare enough! I've made changes to my proposal to avoid CID change pitfall. Now only way to change would be to construct a new block.

Now question is do we have same concern about `BlockView`, note that if it's created from `CID` or a `Block` there is no issue as we can infer both `format` and `cidOpts`. If you do construct fresh `BlockView` (with intention to built up new block / graph) you may not necessarily know the `format` of the block you'd want to create. There for both constructor and `toBlock` take optional  format so that block forma can be chosen on demand. That does however means that one might do

```js
const view = await BlockView.fromCID(c1)
const block = await view.toBlock(otherFormat)
const c1 = await block.toCID()

c1.equals(c2) // => false
```

That being said `format` for `toBlock` is optional and likely won't be passed unnecessarily. However there is the risk, and if you think it's worth addressing I can remove optional param enforce passing `'format'` in the constructor.

**Comment by vmx** [comment]
Thanks @Gozala and @mikeal this is a good read and easy to follow. I agree with the conclusion that a Block should be an immutable CID (can be lazy) + data pair. We need to make sure to document *why* it is like that (as outlined in this issue).

I keep the inline blocks and mutability aside for now. My goal is to find agreement on the basics, where then mutability and inline blocks could be added later on without any (major) changes. One step at a time (that's what I learned from the previous API changes).

Hence I'd leave out the concept of a `BlockView` for now. To me it seems like an IPLD Format with additional functionality. The `BlockView` could be added later on, when we tackle mutability.

So it's pretty much the `Block` that @Gozala proposed with the following changes (based on @mikeals proposal):

 - `fromBytes()` is removed as you would always need to specify a format, hence use the constructor instead.
 - `static toCID()` is removed, use the class method instead.
 - `static toBytes()` is removed, use the `data` property instead.
 - There's `decode()` now, which returns an IPLD Format instance. It kind of replaces the `BlockView.fromBlock()`.
 - There's `encode()` now, which creates a Block out of an object conforming to the IPLD Data Model.
```js
// `IPLDObject` is a JavaScript object that follows the IPLD data model
export opaque type IPLDObject: {} = {}
// `IPLDFormat` is an IPLD Format implementation instance
export opaque type IPLDFormat: Object = Object

declare class Block<a> {
  +data: BinaryEncoded<a>;
  +format: Format;
  equals(mixed): boolean;
  toCID(): Promise<CID<a>>;
  decode(): Promise<IPLDFormat<a>>;

  static constructor(
    data: BinaryEncoded<a>,
    format: Format,
    // Can be passed to override default CID hash algorithm, version for the format.
    options?: CIDOptions
  ): void;
  static encode(IPLDObject<a>, Format, ?CIDOptions): Promise<Block<a>;
}
```

As opposed to @mikeal's proposal, the `encode()` is not lazy. I wonder if it's really needed. What's a use case for creating a block, but not wanting it to be actually serialised? This would make the implementation simpler I guess.

Also there's no `reader()` as in @mikeal's proposal. I consider this part of the IPLD Format stuff and should be figured out there.

*Update*: I've edited the interface quite a bit as I found limitations after posting it. Now I dislike the asymmetry between `encode()` and `decode()`. I hope someone has a better idea.

**Comment by Gozala** [comment]
I've got few pieces of feedback & some inline notes

>  * `fromBytes()` is removed as you would always need to specify a format, hence use the constructor instead.
> * static toBytes() is removed, use the data property instead.

I realize I have not provided much details but toBytes() meant to return `multicodec.addPrefix(this.format, this.data)` and `fromBytes` to do the reverse. It's related to the conversation regarding inline / nested blocks we have going here https://github.com/ipld/ipld/issues/64#issuecomment-477355819 but it makes sense to leave it out and consider as we make more progress on inline / nested blocks.

> * static toCID() is removed, use the class method instead.

I find it really convenient to have static methods / functions doing more functional style stuff, like `blocks.map(Block.toCID)` or doing compositions. It also usually implies that `Block.toCID({data, format})` works without having to construct instance etc.. However if you think it's better to leave it off that's fare as it can be implemented in helper library.

 > - There's decode() now, which returns an IPLD Format instance. It kind of replaces the BlockView.fromBlock().
> - There's encode() now, which creates a Block out of an object conforming to the IPLD Data Model.

This means you're adding whole lot of dependencies here, like codec / format registry. I really wish that can be a separate layer (build on top). That is what `BlockView` was trying to do by staying separate. I understand it also tried to do many other things and maybe that's a point of contention. Maybe `BlockView` could be drastically simplified and be something like:

```js
declare class BlockView<a> {
  toCID(): Eventual<CID<a>>;
  toBlock():Eventual<Block<a>>;
  toBlockData():Eventual<BinaryEncoded<a>>;
  toDagNode(): Eventual<IPLDFormat<a>>;

  static fromBlock(Block<a>):BlockView<a>;
  static fromBlockData(BinaryEncoded<a>, Format, options?:CIDOptions):BlockView<a>;
  static fromDagNode(IPLDObject<a>, Format, ?CIDOptions): Eventual<BlockView<a>>;
}
```

**Comment by Gozala** [comment]
Had some conversation with @vmx and I think I'm going getting tripped over the terminology. I keep thinking of `Block` as being `[bytes, format]` while according to spec [`Block`](https://github.com/ipld/specs#block) is `[bytes, CID]` instead. Which got me thinking about following:

1. Can  `[bytes, CID]` be derived from `[bytes, format]` ? On one hand `hashAlg` can be derived from `format` but then in practice you could create a `CID` that uses a different `hashAlg`. I wonder making that impossible could be a good idea ? In that case `[bytes, CID]` would be just idempotent derivation from `[bytes, format]`.
2. While I really wish `[bytes, format]` was block and there was different term for `[bytes, CID]` I understand that may not be visible at this point in the project life-cycle. But could we at least have a term for `[bytes, format]` ?
    - Maybe term "Block" can be generalized a bit and more concrete terms could be incorporated to refer to an inline / nested blocks (`[bytes, format]`) and CID identified blocks (`[bytes, cid]`). Maybe something like `DataBlock = [bytes, format]` and `NamedBlock = [bytes, cid]`
    - If above option is unreasonable maybe `[bytes, format]` can be something like `EmbeddedBlock`, `InlineBlock`, `AnonymousBlock`, `DataBlock` ?

3. Even though `Block` is term used for `[bytes, CID]` maybe that could be revised to an expansion `[bytes, format, hashAlg]` ? That aligns very well with the idea of lazy CID computation - You can always identify block by computing CID for it, while block itself is details makeing it uniquely addressable.
4. I would also suggest to create a term for encoded block data instead of just referring to it as `Data` which is very broad terms and often fails to deliver desired message. I would offer `BlockBuffer` as it invites analogy with `ArrayBuffer` which conceptually is very similar.
5. I keep getting confused by terms `IPLDFormat` and honestly still not sure what what it's supposed to be. From the comment I get an impression that it's implementation of [IPLD Format API](https://github.com/ipld/interface-ipld-format#api) but `decode(): Promise<IPLDFormat<a>>;` makes me think it's something else.
5. I also find `IPLDObject` to be confusing, can it be referred as `DagNode` or maybe `BlockNode` or maybe `BlockData` ? 
6. We also discussed that maybe `[bytes, format]` should always be paired either via `multicodec.addPrefix(format, bytes)` or as tuple. @vmx suggested that maybe that is what [IPLD Format API](https://github.com/ipld/interface-ipld-format#api) should return on `encode` and take on `decode`. I think it would be a mistake to make API do that because
   - It will be forced to do a thing that can be done generally across all the format implementations, for example once one is added it could be wrapped into something along these lines https://gist.github.com/Gozala/0d653ee628abd15ccfc117605b2c0d6d
   - It creates surface for error where codec `format` doesn't match one being passed `[bytes, format2]`.

**Comment by vmx** [comment]
> * Can  `[bytes, CID]` be derived from `[bytes, format]` ? On one hand `hashAlg` can be derived from `format` but then in practice you could create a `CID` that uses a different `hashAlg`. I wonder making that impossible could be a good idea ? In that case `[bytes, CID]` would be just idempotent derivation from `[bytes, format]`.

To me, being able to use different hashes is a key feature of IPLD.


> * While I really wish `[bytes, format]` was block and there was different term for `[bytes, CID]` I understand that may not be visible at this point in the project life-cycle. But could we at least have a term for `[bytes, format]` ?

Having a term for that would be good.


> * Even though `Block` is term used for `[bytes, CID]` maybe that could be revised to an expansion `[bytes, format, hashAlg]` ? That aligns very well with the idea of lazy CID computation - You can always identify block by computing CID for it, while block itself is details makeing it uniquely addressable.

I think that should be fine. Important thing here is just that there's one specific CID for the block (which could be lazyliy computed).


> * I would also suggest to create a term for encoded block data instead of just referring to it as `Data` which is very broad terms and often fails to deliver desired message. I would offer `BlockBuffer` as it invites analogy with `ArrayBuffer` which conceptually is very similar.

I'd be OK with `BlockBuffer`, though I also don't have a problem with calling it just `data`.


> * I keep getting confused by terms `IPLDFormat` and honestly still not sure what what it's supposed to be. From the comment I get an impression that it's implementation of [IPLD Format API](https://github.com/ipld/interface-ipld-format#api) but `decode(): Promise<IPLDFormat<a>>;` makes me think it's something else.

That is indeed confusing an I'm not happy about `IPLDFormat` myself. The problem here is that it's not clear what and IPLD Format really is.


> * I also find `IPLDObject` to be confusing, can it be referred as `DagNode` or maybe `BlockNode` or maybe `BlockData` ?

I really like `IPLDObject`, it's an JavaScript Object that follows the IPLD Data Model and hence can be encoded with any format that can encode the full IPLD Data Model.


> * We also discussed that maybe `[bytes, format]` should always be paired either via `multicodec.addPrefix(format, bytes)` or as tuple. @vmx suggested that maybe that is what [IPLD Format API](https://github.com/ipld/interface-ipld-format#api) should return on `encode` and take on `decode`. I think it would be a mistake to make API do that because
>   
>   * It will be forced to do a thing that can be done generally across all the format implementations, for example once one is added it could be wrapped into something along these lines https://gist.github.com/Gozala/0d653ee628abd15ccfc117605b2c0d6d

If `encode()` only returns the encoded bytes without the `format`, you would need to keep a reference to the IPLD Format somewhere so that you can put it together someone on a higher level. Instead of keeping that reference "somehow", i'd rather directly return the `format` with the data.

As the higher level is e.g. a Block, which needs `[bytes, format, hash]` you can just use that returned format directly, without the need to read it out directly from `bytes`.

**Comment by achingbrain** [comment]
> Can [bytes, CID] be derived from [bytes, format]

Something to bear in mind is that calculating CIDs is not free.  I would urge you to not throw this information away if you have it, because the first thing a user will do when they notice the absurd CPU usage of their program, is to memoise the CID.

**Comment by mikeal** [comment]
> Something to bear in mind is that calculating CIDs is not free. I would urge you to not throw this information away if you have it, because the first thing a user will do when they notice the absurd CPU usage of their program, is to memoise the CID.

Agreed. This is why I’ve been insistent that we:

* Accept a fully formed CID during instantiation
* Lazily create the CID if not instantiated with one when and only when it is actually required, then obviously cache it if anyone tries to access it again.

**Comment by Gozala** [comment]
Created https://github.com/ipld/ipld/issues/68 so that inline / nested blocks can be discussed separately.

**Comment by mikeal** [comment]
So, this conversation has hit the point where I thought it was better to have some real code.

I implemented the “new stack” (block, codec-interface, and dag-cbor) in this new repo https://github.com/ipld/js-ipld-stack

It has ideas from @Gozala @vmx and @rvagg and myself in it from all the threads we’ve had lately.

For right now I’m doing this all in one repo so that we can test and iterate on the approach without it being spread across several repos but once it’s all ready we’ll just delete the repo and move all the code to the relevant repos.

**Comment by Gozala** [comment]
Seeing above implementation from @mikeal I think I am starting to understand where our views diverge & that is what shows in API choices.

Specifically it appears to me that cases being considered by the new implementation assumes (which is very reasonable) that all of the block is encoded in a same format. While I am personally thinking of cases where block can contain data in multiple formats, that is why in my proposed `BlockView` API there was generic `set` operation: `set<b>(path: string, BlockView<b>): Eventual<BlockView<a>>;` idea is that even if the block itself is in the `dag-cbor` format you could add an entry to it in different format e.g. `dag-pb` or `dag-git`. That is also why I wanted `get<b>(path: string): Eventual<BlockView<b>>;` to return `BlockView` and why  `entries<b>(): AsyncIterable<[string, BlockView<a>]>;` also return `BlockView`s as that would be abstracting over whether entry is a value in the same format, in a different format or a CID.

That also explains why I wanted generic API for Block mutation / composition and why response I was getting is that codec / format should probably provide an API to compose blocks instead (which makes sense when block is fully in that format, but less so when it's not).

**Comment by mikeal** [comment]
@Gozala I’m not actually making that assumption, I’m actually making affordances for inline blocks encoded using an identity multihash. It’s just that, semantically, an inline is *another block instance* and not the *same* block instance.

For most read operations you won’t be using just the single Block reader interface, you’ll be using a multi-block interface. This is where you make affordances for inline blocks to be created out of the identity multihash instead of the getting the CID from a block store. See https://github.com/ipld/js-ipld-stack/blob/master/src/path-level-zero.js#L33

**Comment by rvagg** [comment]
closing for archival
